Fix link resolution for bats

waterkip opened the pull request:
This rewrites the old logic a bit by using `type -p`, `readlink -m`  and
on boxes where we don't have coreutils like `readlink` we use `realpath`
to figure out the correct installation directory of bats.

I've also changes the path of the symbolic link so it works on the
default bash image without having sbin (bats isn't a sbin kinda thing).

Also fixes GH#182

Signed-off-by: Wesley Schwengle <wesley@schwengle.net>

- [x] I have reviewed the [Contributor Guidelines][contributor].
- [x] I have reviewed the [Code of Conduct][coc] and agree to abide by it

[contributor]: https://github.com/bats-core/bats-core/blob/master/docs/CONTRIBUTING.md
[coc]:         https://github.com/bats-core/bats-core/blob/master/docs/CODE_OF_CONDUCT.md

waterkip commented:
You can easily test the bug:
```
docker run -ti --rm --entrypoint /usr/local/bin/bash -v $(pwd):/opt/bats 835f25ff8cf8 bats /opt/bats/test
```

Yields:
```
/usr/sbin/bats: line 44: /libexec/bats-core/bats: No such file or directory
```

This fix:

```
$ docker run -ti --rm --entrypoint /usr/local/bin/bash -v $(pwd):/opt/bats 835f25ff8cf8 bats /opt/bats/test
1..80
ok 1 no arguments prints message and usage instructions
ok 2 invalid option prints message and usage instructions
ok 3 -v and --version print version number
ok 4 -h and --help print help
[snip]
```

pawamoy commented:
Still not merged :confused:

waterkip commented:
@pawamoy it will feel extra good once it is merged. Good things will happen with those who wait ;)

sublimino commented on .gitignore:
```diff
@@ -0,0 +1 @@
+docker-compose.override.yml
```
Is this intentional? `docker-compose` isn't in this repo  : ]

waterkip replied:
I removed the docker bits from this PR and will move it to a different one.

sublimino commented on Dockerfile:
```diff
@@ -7,7 +7,9 @@ FROM bash:${bashver}
 
-RUN ln -s /opt/bats/bin/bats /usr/sbin/bats
-COPY . /opt/bats/
+WORKDIR /opt/bats
```
Is this a breaking change? Users expecting an existing workdir will have to update their use of the Docker container?

waterkip replied:
The default workdir for docker is `/`. It could be breaking if people copy do `COPY . to foo` and expect it to be in `/`. Although I think most people either set the workdir before a copy and than use relative paths or they copy it with the absolute path.

sublimino commented on Dockerfile:
```diff
@@ -7,7 +7,9 @@ FROM bash:${bashver}
-RUN ln -s /opt/bats/bin/bats /usr/sbin/bats
-COPY . /opt/bats/
+WORKDIR /opt/bats
+RUN ln -s /opt/bats/bin/bats /usr/local/bin/bats
```
Another potentially breaking change I think?

waterkip replied:
No, because it is in PATH it isn't a breaking change.

sublimino commented on bin/bats:
```diff
@@ -3,42 +3,70 @@
   BATS_READLINK='readlink'
 fi
 
+set +e
```
Could this be an `if` instead? I *think* this is equivalent

```bash
if ! $BATS_READLINK -m $0 &>/dev/null; then
  BATS_REALPATH='realpath'
fi
```

sublimino commented on bin/bats:
```diff
@@ -3,42 +3,70 @@
+  if [[ -L $path ]]; then
+    # Busybox has no readlink from coreutisl
+    if [[ ! -z "$BATS_REALPATH" ]]; then
+      target_path="$(realpath "$path")"
```
Should this be 

```bash
target_path="$(${BATS_REALPATH} "$path")"
```

?

waterkip replied:
I don't think so, but I'll have a look at the logic when I rework the MR.

sublimino commented on bin/bats:
```diff
@@ -3,42 +3,70 @@
+  elif [[ -f $path ]]; then
+    target_path=$(dirname $path);
+    if [[ ! -z "$BATS_REALPATH" ]]; then
+      target_path="$(realpath "$target_path")"
```
As above, should this use the variable rather than invoking `realpath` directly?

sublimino commented on bin/bats:
```diff
@@ -3,42 +3,70 @@
+set -e
+
+resolve_path() {
+  local path=$1
```
Although variables are case sensitive (*and* this is `local`), it might make sense to rename this for clarity as PATH is obviously used for other things. WDYT?

waterkip replied:
What would be a better naming, it is the path we are talking about. I'm open for suggestions but I personally think, `local` and the complete lack of case-y-ness should make it clear that is isn't the PATH we are talking about especially in the function called `resolve_path`.

sublimino commented on bin/bats:
```diff
@@ -3,42 +3,70 @@
+
 bats_resolve_absolute_root_dir() {
   local cwd="$PWD"
   local path="$1"
```
As above, if there aren't further-reaching implications for this it might be clearer with a different name?

sublimino commented on bin/bats:
```diff
@@ -3,42 +3,70 @@
-  while true; do
-    target_dir="${path%/*}"
-    target_name="${path##*/}"
+  if [[ $path = $(basename $1) ]] ; then
```
Please quote all variables and use `==` for equality in `[[`

sublimino commented on bin/bats:
```diff
@@ -3,42 +3,70 @@
-    target_dir="${path%/*}"
-    target_name="${path##*/}"
+  if [[ $path = $(basename $1) ]] ; then
+    path=$(type -p $path)
```
Quotes around `$path` please

sublimino commented on bin/bats:
```diff
@@ -3,42 +3,70 @@
+  local path=$1
+  local target_path;
+
+  if [[ -L $path ]]; then
```
Please put quotes around `$path`

sublimino commented on bin/bats:
```diff
@@ -3,42 +3,70 @@
+      target_path="$("$BATS_READLINK" -m "$path/../../")"
+    fi
+  elif [[ -f $path ]]; then
+    target_path=$(dirname $path);
```
Please put quotes around `$path`

sublimino commented on bin/bats:
```diff
@@ -3,42 +3,70 @@
-      return
-    fi
-  done
+  target_path="$(resolve_path $path)"
```
Please put quotes around `$path`

sublimino commented on bin/bats:
```diff
@@ -3,42 +3,70 @@
-    fi
-  done
+  target_path="$(resolve_path $path)"
+  if [[ $target_path = '.' ]];
```
Quotes around `$target_path` please

sublimino commented on bin/bats:
```diff
@@ -3,42 +3,70 @@
+  target_path="$(resolve_path $path)"
+  if [[ $target_path = '.' ]];
+  then
+    target_path="$(resolve_path $target_path)"
```
Quotes around `$target_path` please

sublimino commented on docker-compose.override.dist:
```diff
@@ -0,0 +1,8 @@
+# Copy this file to docker-compose.override.yml
```
Could we move the `docker-compose` work out into another PR please? It needs some tests around it too please : ]

sublimino commented on test/root.bats:
```diff
@@ -40,28 +40,45 @@ setup() {
+@test "set BATS_ROOT from PATH" {
+  cd "$BATS_TEST_SUITE_TMPDIR"
+  cd /tmp
+  PATH=$PATH:$BATS_TEST_SUITE_TMPDIR/bin
```
Quotes around the variable assignment please

waterkip commented:
> Sorry, I'm so busy it's difficult to find time, but better late than never I hope 
> 
> Some nits here, mostly quoting, and some possible breaking changes in the Dockerfile I'd prefer to avoid if we can, or add something to the release notes otherwise.
> 
> I'm still intending to release the next version before merging changes, but this looks good to go when that's done, thanks for the hard work on this!

I moved the docker bits to a different PR and updated the PR to reflect your comments. Some bits could be left out I found out during an autosquash. If you could have another look that would be awesome :)
